### DISCUSSION TITLE
Use context with getStaticProps to avoid prop drilling

### DISCUSSION THREAD

**Discussion by wiesson** [vercel/next.js#17420] [Help] [answered]
I'm trying to populate a Context (React.Context) with language (static) data from getStaticProps to avoid prop-drilling. Is this supposed to work like this? 

Example

**app.jsx**
```Javascript
export const TranslationContext = React.createContext({}); // Somehow I'd like to fill the initialContext with data from getStaticProps

function MyApp({ Component, pageProps }) {
  const { translations, language, ...rest } = pageProps;
  // translation and language is set during SSR
  return (
    <TranslationContext.Provider
      value={{ translations, language }}
    >
      <Component {...rest} />
    </TranslationContext.Provider>
  );
}
```

**some/deeply/nested/component/Title.jsx**
```Javascript
const Title = () => {
  const context = React.useContext(TranslationContext);
  return <div>{context.translations.MY_HEADLINE}</div>
}
export default Title;
```

I have the feeling that the context will be populated later in the client and not during SSR. Any ideas?

**Comment by jamesmosier** [comment, marked as answer]
The Context API should work as you intend with your example with `getStaticProps` (and `getServerSideProps` for that matter).

You'll just have to make sure each page has a `translations` and `language` prop so that your context provider gets its value since there is no global way to fetch data in `_app.js` (aside from `getInitialProps` which is not recommeded). 

Also it would probably be best to export your `export const TranslationContext = React.createContext({});` from a file outside of the `pages` directory.

**Reply by Goncaloo42** [reply]
I tried to reproduce what you explained but i cannot get the expected result. I have a page with the getStaticProps call, in that page, i can use the context API value with no problems.. but if i try in another page (that doesn't implement getStaticProps), the object that i get is undefined...

**Reply by SSylvain1989** [reply]
did you find a solution @Goncaloo42

**Comment by wiesson** [comment]
Thanks!

I have recreated a minimal example with a fresh Nextjs installation and it worked out of the box, so I deleted the `node_modules` folder in my main project and did a fresh `yarn install` and it works 🤷 thanks 😅

**Comment by wiesson** [comment]
~~Ok, I found the issue which may caused problems - turns out, with webpack5, the Context API does not work as expected during SSR. I have made a minimalistic example:~~

~~https://github.com/wiesson/nextjs-context-example
https://codesandbox.io/s/clever-pond-r2r2e~~

Can't reproduce it anymore 🤷 - maybe fixed with webpack5 rc1 or I had some other local issue
